### PR TITLE
Make player liquid speed independent of FPS

### LINKED ISSUE #7530: Player movement speed in liquids is FPS-dependent
##### Issue type

- Bug report
 
##### Minetest version

```
0.4.17
```

##### OS / Hardware

Operating system: Fedora 28

##### Summary

The higher the client FPS is, the slower the player will swim in liquids; this is especially noticeable with water in minetest_game. This puts players with fast machines and high framerates at a disadvantage in multiplayer.

(I play with the FPS cap disabled to [decrease input latency](http://danluu.com/input-lag/), which should be supported by any modern game.)

##### Steps to reproduce

1. Tweak settings so that you can reach high FPS values (values above 150 make it easier to notice this issue).

2. Disable the FPS cap in minetest.conf using this setting:

```
fps_max = 0
```

3. Run the game (singleplayer or multiplayer) and swim in water.

### PR DESCRIPTION
**Pull request by paramat**
2nd commit: Fix codestyle issues in code block 
//////////////////////
Closes #7530 

The code in ClientEnvironment::step(float dtime) splits 'dtime' (1 / FPS)  into several smaller 'dtime_part's, each part being a maximum 0.0100s.
With my FPS at 60, printing 'dtime_part' gives:
```
  0.0100
  0.0060
  0.0100
  0.0070
  0.0100
  0.0060
  0.0100
  0.0070
  0.0100
  0.0070
  0.0100
  0.0070
  0.0100
  0.0060
  0.0100
  0.0070
  0.0100
  0.0060
  0.0100
  0.0070
  0.0100
  0.0070
```
Each 'dtime' of 1/60 = 0.0167s is split into 2 parts.

With my FPS at 30, printing 'dtime_part' gives:
```
  0.0100
  0.0100
  0.0100
  0.0030
  0.0100
  0.0100
  0.0100
  0.0030
  0.0100
  0.0100
  0.0100
  0.0030
  0.0100
  0.0100
  0.0100
  0.0030
  0.0100
  0.0100
  0.0100
  0.0030
```
Each 'dtime' of 1/30 = 0.033s is split into 4 parts.

Every 'dtime_part', speed is reduced to slow the player in liquid.

For very low FPS 'dtime' is split into many parts of 0.0100 plus 1 variable part. So 'dtime_part' averages roughly 0.0100, so speed is reduced roughly 100 times per second.
At 99FPS 'dtime' is 0.0101 which is split into 2 parts 0.0100 + 0.001 so 'dtime_part' averages roughly 0.0050, so speed is reduced roughly 200 times per second.
At 101FPS 'dtime' is 0.0099 which is split into one part 0.0099 so 'dtime_part' averages to 0.0099,  so speed is reduced 101 times per second.

This must be the cause of liquid speed depending on FPS. There should be a big change between 99 and 101 FPS.

This is not straightforward to fix then and the current PR code is being reconsidered.

### CODE REVIEW

**Comment by paramat** [comment]
For most player speeds 'dtime_max_increment' = 0.0100.
So perhaps:
`v3f d = d_wanted.normalize() * (dl * dtime_part / dtime_max_increment);`
When 'dtime_part' = 0.0100, 'dtime_part / dtime_max_increment' will be 1.
But when 'dtime_part' is smaller, that will then be taken into account and
'dtime_part / dtime_max_increment' will fall below 1.

**Comment by paramat** [comment]
Updated with that, it seems to work for FPS 10, 30 and 60.
@Calinou could you test at higher FPS? Especially looking for any change between 99 and 101 FPS?

**Comment by paramat** [comment]
Now using `* 100.0f` instead of `/ dtime_max_increment`:
`v3f d = d_wanted.normalize() * (dl * dtime_part * 100.0f);`
Because this is a deceleration, so should always be proportional to 'dtime_part' what ever value is chosen for `dtime_max_increment`.
Speed in liquid will change slightly depending on your FPS, because it is now independent of FPS. Those playing at  60FPS will notice a slight increase in liquid speed.

File: src/clientenvironment.cpp [hunk 1]
```diff
@@ -165,31 +165,32 @@ void ClientEnvironment::step(float dtime)
...
-				if(lplayer->in_liquid_stable || lplayer->in_liquid)
-				{
-					// How much the node's viscosity blocks movement, ranges between 0 and 1
-					// Should match the scale at which viscosity increase affects other liquid attributes
-					const f32 viscosity_factor = 0.3;
+				if (lplayer->in_liquid_stable || lplayer->in_liquid) {
+					// How much the node's viscosity blocks movement, ranges
+					// between 0 and 1. Should match the scale at which viscosity
+					// increase affects other liquid attributes.
+					const f32 viscosity_factor = 0.3f;
```
Comment (paramat) [inline src/clientenvironment.cpp]: Should this be a 'static const'?

File: src/clientenvironment.cpp [hunk 2]
```diff
@@ -165,31 +165,32 @@ void ClientEnvironment::step(float dtime)
...
 
 					v3f d_wanted = -speed / lplayer->movement_liquid_fluidity;
 					f32 dl = d_wanted.getLength();
```
Comment (paramat) [inline src/clientenvironment.cpp]: Should i change 'f32' to 'float'?

File: src/clientenvironment.cpp (hunk 1 above)
Comment (paramat) [reply src/clientenvironment.cpp]: Did it anyway. It's run 100 times a second.

**Comment by ClobberXD** [comment]
Tested with 20 and 60 FPS. Duration taken to swim from point A -> B was the same in both cases. :+1:

**Pull request merged**